src/envoy/utils/utils.cc

```diff
@@ -178,6 +178,11 @@ bool IsMutualTLS(const Network::Connection* connection) {
          connection->ssl()->peerCertificatePresented();
 }
 
+bool IsUpstreamMutualTLS(const StreamInfo::StreamInfo& stream_info) {
+  const auto ssl = stream_info.upstreamSslConnection();
+  return ssl != nullptr && ssl->peerCertificatePresented();
```

Comment:
I'm a little confused on the usage of `ssl->peerCertificatePresented()`. If this flag is set to true, does this imply client was using TLS for talking to upstream and got a successful certificate from upstream or it should mean that client is using TLS irrespective of how the negotiation went and what upstream server presented?

Reply:
Ouch, I think this might be an issue. We should be checking whether a local certificate was presented, not peer.

Reply:
that's what I thought. I don't see an Envoy API for that. @lizan may be you know better?

Reply:
We can extend envoy to report whether a local certificate is presented. That might not mean that the server validated it. Not sure how to get that info.

Reply:
we didn't have this in Envoy API because this has been simply reflecting what in the TLS context config, though we're adding more dynamic nature to the config for incremental adoption etc, so it makes sense to add, thanks for posting https://github.com/envoyproxy/envoy/pull/8464.